text/0000-singleton-removal.md

```diff
@@ -0,0 +1,69 @@
+- integration testing. Both games that use MLAPI and MLAPI itself can write full server/client tests that run entirely in the scope of a single process. 
+- clients talking to multiple servers simultaneously. This is expected to be a requirement for MMO support. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+MLAPI supports having multiple `NetworkingManager`s running concurrently. There are a few things to keep in mind when doing so:
+- The simplest case is to only have one `NetworkingManager` per scene. This is what you do to support an in-process server: load a "server" version of your scene, and then call `StartServer` on the `NetworkingManager` in it. All "server" physics calculations must then be performed via the `PhysicsScene` associated with the server scene.
+- If you have multiple `NetworkingManager`s in one scene (for example, to talk to two different servers at once), then you should take care that all placed `NetworkedObject`s be wired to the `NetworkingManager` they are associated with via a serialized field.
+- `NetworkedObject`s spawned by a `NetworkingManager` will always be bound to the `NetworkingManager` that created them, and be placed in the same scene as their parent `NetworkingManager`.
+- If launching an in-process server, be sure to only call `StartClient` after the `OnServerStarted` callback has fired. 
+- Currently only the ENet Transport is supported. UNet shares a common static at the Unity level (`NetworkTransport`), which prevents it from being fully separated.
+- We have removed the old `NetworkingManager.Singleton` method. For migrators, if you are writing code that in a component sitting on a `NetworkedObject`, you can simply get the `NetworkedObject`s `NetManager` instead. For other code, you will need to adopt your own strategy (for example, finding the `NetworkingManager` in the same scene as yourself via [`GameObject.FindWithTag`](https://docs.unity3d.com/ScriptReference/GameObject.FindWithTag.html)).
+- Note that the MLAPI Profiler doesn't support profiling multiple `NetworkingManager`s simultaneously. If run when multiple `NetworkingManager`s are active, it will profile the first `NetworkingManager` it finds.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`NetworkingManager.Singleton` is removed. `SpawnManager`, `InternalMessageSender`, `InternalMessageHandler`, `NetworkSceneManager`, and `BitReaderPool` have all been made non-static, and hang off the `NetworkingManager` that owns them. 
+
+`NetworkedObject`s have a `NetManager` field that is stamped on them at their moment of creation (when spawned by MLAPI), or should already be set in editor (for placed objects). If still unset by the time `NetworkedObject.Start` runs, the `NetworkedObject` takes the first `NetworkingManager` it finds in the same scene as itself.
```

Comment:
I think this implicit behavior is somewhat shaky IMHO. I wouldn't expect `NetworkManager` or `NetworkObject` to find each other automatically unless set explicitly. There might be a good chance to create another `NetworkLinker` or something similar that has multiple linking options which could link `NetworkObject`/`NetworkBehaviour` to `NetworkManager`. In one of the modes, `NetworkLinker` could find all unregistered `NetworkObject`s and register them with the first `NetworkManager` instance in the scene. Or maybe `NetworkLinker` could be further extended in the future where is routes registration of `NetworkObject` based on some sort of condition and maps it to a specific `NetworkManager` instance. To put more simply, I think outsourcing this particular need to another 3rd `NetworkLinker` component would be a better approach except linking them by hand explicitly. What do you think? :)

Reply:
I think I could live with this.  I was tempted to suggest it assert loud and proud rather than magically fixing the NetworkingManager (NM), but then I think "when won't I want an object I instantiated in the scene to not be managed by that scene".

I am curious though how you think one would set the NM of an object in the editor?  Would the NMs be enumerable to where you could see a pickable list in the editor?  Or would the setting of the NM need to be in code?

Reply:
Currently, `NetworkManager` is a `MonoBehaviour` which means it's being serialized in Editor window therefore one could set a `NetworkObject`'s `NetworkManager` field in the Editor via inspector window. We could even script `NetworkObject` editor window entirely to give it a completely customized look in the inspector. There are lots of ways to achieve that in the Editor.

But having said that, this work is not required nor essential so we could address object-manager linking in the future and keep this auto-register idea around for now. I do not want to extend the scope here.

Reply:
My approach changed a little bit when I actually came to implement this. Now it's the NetworkManager that stamps itself on the NetworkObject (via `NetworkSpawnManager.ServerSpawnSceneObjectsOnStartSweep` and the equivalent path on the client). 

Importantly, the NetworkManager won't stamp itself on a NetworkObject that is already claimed by a NetworkManager. Right now, this creates a worrying order-dependency for startup. If you have a "Chat" NetworkManager and a "Game" NetworkManager, you better hope the "Game" one starts up first!

I think a good enhancement would be a `NetworkConfig.IgnoreSceneObjects` option that you could set to true. That way you could enforce that your Chat NetworkManager never steals your scene objects.